### PR TITLE
Support optimization of replicated table queries

### PR DESCRIPTION
**Pull request by asubramanya**
This commit adds support for optimization of SELECT and INSERT queries
involving replicated tables. UPDATE and DELETE are currently not supported.
The commit introduces a new distribution policy to represent replicated tables.
This new distribution policy is mapped to the existing distribution spec
`CDistributionSpecReplicated`.

In order to optimize queries for replicated tables, we need to account for
replicated distribution when setting Input segments for Gather motion. If the
tree underneath of Physical Gather Motion has a distribution spec as
"Replicated" then, the number of input segments for Gather are set to 1 by
`FDuplicateHazardMotion()`. However `CheckValidity()` did not allow Gather
Motion with only one Input Segments. Now with replicated tables, there will be
scenarios where Gather needs to happen with a single segment as input. Also the
original commit which introduced the function added a test, but it no longer
reproduces the error. So `CheckValidity()` has been removed.

Co-authored-by: Abhijit Subramanya <asubramanya@pivotal.io>

### CODE REVIEW

File: libgpopt/include/gpopt/base/COptCtxt.h [hunk 1]
```diff
@@ -99,6 +99,9 @@ namespace gpopt
 			// functions that read/modify SQL data
 			BOOL m_has_volatile_or_SQL_func;
 
+			// if there are replicated tables in the query
```
Comment (vraghavan78) [inline libgpopt/include/gpopt/base/COptCtxt.h]: does the query have replicated tables

**Comment by dhanashreek89** [comment]
>> select and insert involving replicated tables are supported 

Is the INSERT on a replicated table supported as well ? I believe yes ?

**Comment by asubramanya** [comment]
@dhanashreek89 Yes. I have added a simple minidump test as well.

File: data/dxl/minidump/ReplicatedTableInsert.mdp [hunk 2]
```diff
@@ -0,0 +1,236 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<dxl:DXLMessage xmlns:dxl="http://greenplum.com/dxl/2010/12/">
+  <dxl:Comment><![CDATA[
+	 CREATE TABLE r1(a INT, b INT);
```
Comment (hardikar) [inline data/dxl/minidump/ReplicatedTableInsert.mdp]: Can we be explicit here that its `DISTRIBUTED REPLICATED;`? We might change defaults anytime :)

Comment (asubramanya) [reply data/dxl/minidump/ReplicatedTableInsert.mdp]: Yes. Will include that.

**Comment by asubramanya** [comment]
@hardikar Thanks for the suggestions. The corresponding GPDB side PR: https://github.com/greenplum-db/gpdb/pull/5912

File: libgpopt/src/translate/CTranslatorExprToDXL.cpp [hunk 3]
```diff
@@ -3951,33 +3951,6 @@ CTranslatorExprToDXL::PdxlnHashJoin
...
 
-//---------------------------------------------------------------------------
-//	@function:
-//		CTranslatorExprToDXL::CheckValidity
-//
-//	@doc:
-//		Check if the motion node is valid
-//---------------------------------------------------------------------------
-void
-CTranslatorExprToDXL::CheckValidity
```
Comment (bhuvnesh2703) [inline libgpopt/src/translate/CTranslatorExprToDXL.cpp]: Do we have cases other than on replicated tables where the plan with the below inconsistency were marked invalid?

**Comment by hardikar** [comment]
Nice work! Thanks for adding all the tests! I'll will definitely help us a lot down the line.

**Pull request merged**